chips/msp432/src/spi.rs

```diff
@@ -0,0 +1,476 @@
+                as *const ReadWrite<u16, usci::UCSPIxCTLW0::Register>)
+        }
+    }
+
+    fn brw(&self) -> &ReadWrite<u16> {
+        &self.brw
+    }
+
+    fn statw(&self) -> &ReadWrite<u16, usci::UCSPIxSTATW::Register> {
+        unsafe {
+            &*(&self.statw as *const ReadWrite<u16, usci::UCAxSTATW::Register>
+                as *const ReadWrite<u16, usci::UCSPIxSTATW::Register>)
+        }
+    }
+
+    fn rxbuf(&self) -> &ReadOnly<u16> {
+        &self.rxbuf
+    }
+
+    fn txbuf(&self) -> &ReadWrite<u16> {
+        &self.txbuf
+    }
+
+    fn ie(&self) -> &ReadWrite<u16, usci::UCSPIxIE::Register> {
+        unsafe {
+            &*(&self.ie as *const ReadWrite<u16, usci::UCAxIE::Register>
+                as *const ReadWrite<u16, usci::UCSPIxIE::Register>)
+        }
+    }
+
+    fn ifg(&self) -> &ReadWrite<u16, usci::UCSPIxIFG::Register> {
+        unsafe {
+            &*(&self.ifg as *const ReadWrite<u16, usci::UCAxIFG::Register>
+                as *const ReadWrite<u16, usci::UCSPIxIFG::Register>)
+        }
+    }
+}
+
+#[derive(Clone, Copy, PartialEq)]
+enum OperatingMode {
+    Unconfigured,
+    Idle,
+    Write,
+    WriteRead,
+}
+
+#[derive(Clone, Copy)]
+#[repr(u16)]
+enum SpiClock {
+    K1500 = 1, // 1.5MHz
+    K750 = 2,  // 750kHz
+    K500 = 3,  // 500kHz
+    K375 = 4,  // 375kHz
+    K300 = 5,  // 300kHz
+    K250 = 6,  // 250kHz
+    K150 = 10, // 150kHz
+    K100 = 15, // 100kHz
+}
+
+impl From<u32> for SpiClock {
+    fn from(rate: u32) -> Self {
+        match rate {
+            0..=124_999 => SpiClock::K100,
+            125_000..=199_999 => SpiClock::K150,
+            200_000..=274_999 => SpiClock::K250,
+            275_000..=337_499 => SpiClock::K300,
+            337_500..=437_499 => SpiClock::K375,
+            437_500..=624_999 => SpiClock::K500,
+            625_000..=1124_999 => SpiClock::K750,
+            _ => SpiClock::K1500,
+        }
+    }
+}
+
+impl From<SpiClock> for u32 {
+    fn from(clk: SpiClock) -> Self {
+        match clk {
+            SpiClock::K100 => 100_000,
+            SpiClock::K150 => 150_000,
+            SpiClock::K250 => 250_000,
+            SpiClock::K300 => 300_000,
+            SpiClock::K375 => 375_000,
+            SpiClock::K500 => 500_000,
+            SpiClock::K750 => 750_000,
+            SpiClock::K1500 => 1_500_000,
+        }
+    }
+}
+
+pub struct Spi<'a> {
+    registers: &'static dyn UsciSpiRef,
+    cs: OptionalCell<&'a dyn Pin>,
+    hold_cs: Cell<bool>,
+    operating_mode: Cell<OperatingMode>,
+    clock: Cell<SpiClock>,
+    tx_buf: TakeCell<'static, [u8]>,
+    master_client: OptionalCell<&'a dyn spi::SpiMasterClient>,
+
+    tx_dma: OptionalCell<&'a dma::DmaChannel<'a>>,
+    pub(crate) tx_dma_chan: usize,
+    tx_dma_src: u8,
+
+    rx_dma: OptionalCell<&'a dma::DmaChannel<'a>>,
+    pub(crate) rx_dma_chan: usize,
+    rx_dma_src: u8,
+}
+
+impl<'a> Spi<'a> {
+    pub fn new(
+        registers: &'static dyn UsciSpiRef,
```

Comment:
I'd suggest parameterizing Spi by the `UsciSpiRef` implementation rather than having a `dyn`.